### PR TITLE
Check assertion and assume incrementally using the push-pop functions

### PR DESCRIPTION
**Pull request by lucasccordeiro**
This PR allows us to check each assertion and assume incrementally using the push-pop functions available in the SMT solvers. If the assertion or assume statement turns to be true, we remove it from the resulting logical formula to be passed to the SMT solver; otherwise, we negate the assertion (for assertion) and add an assume(false) to cut the subsequent execution paths.

### CODE REVIEW

File: src/goto-symex/symex_main.cpp [hunk 1]
```diff
@@ -54,6 +54,30 @@ void goto_symext::claim(const expr2tc &claim_expr, const std::string &msg)
...
+    auto rte = std::dynamic_pointer_cast<runtime_encoded_equationt>(target);
+    equality2tc question(gen_true_expr(), new_expr);
+    try
+    {
+      // check whether the assertion holds
+      tvt res = rte->ask_solver_question(question);
+      if(res.is_true())
+        // we don't add this assertion to the resulting VCs
+        return;
+      else if(res.is_false())
```
Comment (mikhailramalho) [inline src/goto-symex/symex_main.cpp]: No need for an `else` here.

File: src/goto-symex/symex_main.cpp [hunk 2]
```diff
@@ -54,6 +54,30 @@ void goto_symext::claim(const expr2tc &claim_expr, const std::string &msg)
...
+      {
+        // negate the claim and add it as an assumption
+        not2tc not_new_expr(new_expr);
+        assume(not_new_expr);
```
Comment (mikhailramalho) [inline src/goto-symex/symex_main.cpp]: I don't think we need this here: we can convert it to an `assume(false)`.

Also, I'd be interested in halting the symex here: we know this is a assert violation, so we can save a lot of time by stopping and printing the cex.

File: src/goto-symex/symex_main.cpp (hunk 1 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_main.cpp]: Fixed.

File: src/goto-symex/symex_main.cpp [hunk 3]
```diff
@@ -54,6 +54,31 @@ void goto_symext::claim(const expr2tc &claim_expr, const std::string &msg)
...
+        return;
+      else if(res.is_false())
+        // this assertion is violated
+        is_violated = true;
+      else
+        this->msg.status("Incremental verification returned unknown");
+    }
+    catch(runtime_encoded_equationt::dual_unsat_exception &e)
+    {
+      assert(0 && "Can't check this assertion, sorry");
```
Comment (rafaelsamenezes) [inline src/goto-symex/symex_main.cpp]: Who throws this exception? Does it give more info? If so, try to explain the reason why it was throw.

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]

**Review by rafaelsamenezes — APPROVED** [review]

**Review by mikhailramalho — CHANGES_REQUESTED** [review]
This is more convoluted than it has to be, how about:
1. ask question
2. if it's true, return an do nothing
3. if it's false, add assume(false) and return
4. Otherwise, print message saying that we found a constraint that cannot be solved and we'll fall back to solve the whole equation.

Other comments:
1. No need for else if the previous if-block has a return.
2. Do we need to abort when we catch a dual_unsat_exception exception?

File: src/goto-symex/symex_main.cpp [hunk 4]
```diff
@@ -54,6 +93,13 @@ void goto_symext::claim(const expr2tc &claim_expr, const std::string &msg)
...
+  if(options.get_bool_option("smt-symex-assert"))
+  {
+    if(check_incremental(new_expr, true, msg))
+      // incremental verification has succeeded
+      return;
+  }
+
   cur_state->guard.guard_expr(new_expr);
   cur_state->global_guard.guard_expr(new_expr);
   remaining_claims++;
```
Comment (mikhailramalho) [inline src/goto-symex/symex_main.cpp]: Could use also replace this code with a call to the new function `assertion`?

File: src/goto-symex/symex_main.cpp [hunk 5]
```diff
@@ -25,6 +25,45 @@
...
+    return false;
+  }
+  catch(runtime_encoded_equationt::dual_unsat_exception &e)
+  {
+    this->msg.error(
+      "This solver was unable to check this expression. Please try it with "
+      "another solver");
+  }
+  return false;
+}
```
Comment (mikhailramalho) [inline src/goto-symex/symex_main.cpp:61]: add a newline here.

File: src/goto-symex/symex_main.cpp [hunk 6]
```diff
@@ -313,6 +373,12 @@ void goto_symext::symex_assume()
   dereference(cond, dereferencet::READ);
   replace_dynamic_allocation(cond);
 
+  if(options.get_bool_option("smt-symex-assume"))
+  {
+    if(check_incremental(cond, false, std::string()))
+      // incremental verification has succeeded
+      return;
+  }
```
Comment (mikhailramalho) [inline src/goto-symex/symex_main.cpp]: Is this correct? We should return when the early if the assume is true, but `check_incremental` also returns true when cond evaluates to false.

You'll probably need to replace the return value with a threeval object or an optional.

File: src/goto-symex/symex_main.cpp [hunk 7]
```diff
@@ -25,6 +25,45 @@
 #include <util/std_expr.h>
 #include <vector>
 
+bool goto_symext::check_incremental(
+  expr2tc &expr,
```
Comment (mikhailramalho) [inline src/goto-symex/symex_main.cpp]: expr can be const here: when calling assertion and assume, you can use gen_false_expr() directly.

File: src/goto-symex/symex_main.cpp [hunk 8]
```diff
@@ -54,6 +93,13 @@ void goto_symext::claim(const expr2tc &claim_expr, const std::string &msg)
   if(is_true(new_expr))
     return;
 
+  if(options.get_bool_option("smt-symex-assert"))
+  {
+    if(check_incremental(new_expr, true, msg))
+      // incremental verification has succeeded
+      return;
+  }
+
```
Comment (mikhailramalho) [inline src/goto-symex/symex_main.cpp:98]: Shouldn't we add the result of the evaluation to the guards if the assertion evaluates to false?

**Review by mikhailramalho — CHANGES_REQUESTED** [review]

File: src/goto-symex/symex_main.cpp (hunk 4 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_main.cpp]: Done.

File: src/goto-symex/symex_main.cpp (hunk 5 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_main.cpp:61]: Fixed.

File: src/goto-symex/symex_main.cpp (hunk 6 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_main.cpp]: The method `check_incremental` already adds `assume(false)`.

File: src/goto-symex/symex_main.cpp (hunk 7 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_main.cpp]: Fixed.

File: src/goto-symex/symex_main.cpp (hunk 8 above)
Comment (lucasccordeiro) [reply src/goto-symex/symex_main.cpp:98]: Fixed.

**Review by mikhailramalho — APPROVED** [review]
LGTM

**Pull request merged**